examples/dominant-assurance/index.rsh

```diff
@@ -0,0 +1,134 @@
+  entrepreneurInvestment: UInt, // Entrepreneur's contribution to the product funding
+  entrepreneurProfit: UInt,     // Entrepreneur's profit when the quorum is met
+  investorInvestment: UInt,     // Each investor's contribution to the product funding
+  investorFailProfit: UInt,     // Each investor's profit if the quorum is not met
+  investorQuorum: UInt,         // Number of investors needed to successfully fund the product
+  investmentDuration: UInt,     // How long funding will be open to investors
+  failPayDuration: UInt,        // How long failure pay will be available for investors to claim
+});
+
+export const main = Reach.App(() => {
+  const P = Participant('Product', {
+    investmentStructure: InvestmentStructureT,
+    ready: Fun([], Null),
+  });
+  const E = Participant('Entrepreneur', {});
+  const I = API('Investor', {
+    invest: Fun([], Null),
+    collectFailPay: Fun([], Null),
+  });
+  setOptions({ untrustworthyMaps: true });
+  init();
+
+  const checkInvestmentStructure = (iso) => {
+    const expectedFunds = iso.entrepreneurInvestment
+                        + iso.investorQuorum * iso.investorInvestment;
+    check(iso.entrepreneurProfit <= expectedFunds);
+    check(iso.investorQuorum > 1);
+    check(iso.investorInvestment > 0);
+    check(iso.investorFailProfit > 0);
+  };
+
+  // A participant representing the product being funded specifies how investment will work
+  P.only(() => {
+    const investmentStructure = declassify(interact.investmentStructure);
+    checkInvestmentStructure(investmentStructure);
+  });
+  P.publish(investmentStructure);
+  checkInvestmentStructure(investmentStructure);
+
+  const {
+    entrepreneurInvestment,
+    entrepreneurProfit,
+    investorInvestment,
+    investorQuorum,
+    investorFailProfit,
+    investmentDuration,
+    failPayDuration
+  } = investmentStructure;
+
+  commit();
+
+  // Entrepreneur kicks off the contract by paying their investment plus enough
+  // to compensate investors in the case of failure
+  const starterInvestment = entrepreneurInvestment
+                          + investorFailProfit * (investorQuorum - 1);
+  E.pay(starterInvestment);
+  P.interact.ready();
+
+  const investmentTimeout = relativeTime(investmentDuration);
+  const investors = new Set();
+
+  // Investors are given a change to invest
+  const [timedOut, numInvestors] =
+    parallelReduce([false, 0])
+    .invariant(
+      balance() == starterInvestment + numInvestors * investorInvestment
+      // && investors.Map.size() == numInvestors
+      && numInvestors <= investorQuorum
+    )
+    .while(!timedOut && numInvestors < investorQuorum)
+    .api(I.invest,
+      () => check(!investors.member(this)),
+      () => investorInvestment,
+      (k) => {
+        check(!investors.member(this));
+        investors.insert(this);
+        k(null);
+        return [false, numInvestors + 1];
+      }
+    )
+    .timeout(investmentTimeout, () => {
+      Anybody.publish();
```

Comment:
For a real program, this should really be an API call because we can't expect a participant to be attached the whole time